### PR TITLE
Import delegations via CSV

### PR DESCRIPTION
**Pull request by laurajaime**
This PR add importation button to import delegations via CSV.

The csv has this format:
`from, to`

### Testing
1. Go to Participants
2. Delegations
3. Select a consultation
4. In delegations list > Import via csv
5. See new delegations in list

### Screenshots
imagen

### CODE REVIEW

File: app/controllers/decidim/action_delegator/admin/delegations_csv_imports_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        helper DelegationHelper
+        helper_method :current_setting
+
+        layout "decidim/admin/users"
+
+        def new
+          enforce_permission_to :csv_import, :setting
+
+          @form = form(DelegationsCsvImportForm).instance
+          render template: "decidim/action_delegator/admin/settings/csv_import/new"
```
Comment (tramuntanal) [inline app/controllers/decidim/action_delegator/admin/delegations_csv_imports_controller.rb:16]: Why all the controller is not under Settings?

File: app/controllers/decidim/action_delegator/admin/delegations_csv_imports_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+          ImportDelegationsCsv.call(@form, current_user, current_setting) do
+            on(:ok) do
+              flash[:notice] = I18n.t(".delegations.csv_imports.success", scope: "decidim.action_delegator.admin")
+              redirect_to setting_delegations_path(current_setting)
+            end
+
+            on(:invalid) do
+              flash[:alert] = I18n.t(".delegations.csv_imports.invalid", scope: "decidim.action_delegator_admin")
+              render template: "decidim/admin/delegations_csv_imports/new"
```
Comment (tramuntanal) [inline app/controllers/decidim/action_delegator/admin/delegations_csv_imports_controller.rb:31]: is this path correct? shoudn't it be the same as in the `new` action?

**Review by microstudi — COMMENTED** [review]
Thanks for this @laurajaime 

However, we are working on a refactored codebase of this plugin. However, this feature should be pretty much straightforward to adapt.
Can you redirect this PR to the `develop` branch?

Also, see if this #139 helps you in any way, as it does more or leas the same thing for participants (also including a nice email report with the result of the import). It would be nice to have the same sort of behavior of this one.

Note that the "develop" branch will become the "main" branch very soon (probably next week). I don't think we will merge anything in the main branch before that.

**Review by tramuntanal — COMMENTED** [review]
Good job  @laurajaime !
I've left you some questions that I need you tu clarify please :smile:

**Comment by microstudi** [comment]
@tramuntanal please take a look at the "develop" branch, we have almost completed the refactoring for this plugin!
¡

**Comment by laurajaime** [comment]
> Thanks for this @laurajaime
> 
> However, we are working on a refactored codebase of this plugin. Still this feature should be pretty much straightforward to adapt. Can you redirect this PR to the `develop` branch?
> 
> Also, see if this #139 helps you in any way, as it does more or leas the same thing for participants (also including a nice email report with the result of the import). It would be nice to have the same sort of behavior of this one.
> 
> Note that the "develop" branch will become the "main" branch very soon (probably next week). I don't think we will merge anything in the main branch before that.

Hi @microstudi , I'm going to make the change based on the PR https://github.com/coopdevs/decidim-module-action_delegator/pull/139.

**Comment by laurajaime** [comment]
@microstudi can you review it now, please? Thanks you!!

**Comment by microstudi** [comment]
right away! thanks @laurajaime

File: app/views/decidim/action_delegator/admin/manage_delegations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      <h2 class="card-title">
+        <%= t(".title") %>
+      </h2>
+    </div>
+  </div>
+  <p><%= t(".upload_instructions") %></p>
+  <p><%= t(".required_fields") %></p>
+  <p><%= t(".title_example") %></p>
+  <pre class="code-block">
+    granter_email,grantee_email
```
Comment (microstudi) [inline app/views/decidim/action_delegator/admin/manage_delegations/new.html.erb:14]:
```suggestion
    from,to
```
I have the feeling that the words `granter` and `grantee` are very confusing. I would like to know your opinion on this. Do you thing that something simpler like `from`, `to` would simplify this?

**Comment by antopalidi** [comment]
Good job @laurajaime 
In my opinion this can be improved by creating a concern to reuse the same code.
The delegations import and participants import are identical, but each with its own data.

File: app/views/decidim/action_delegator/admin/manage_delegations/new.html.erb (hunk 3 above)
Comment (laurajaime) [reply app/views/decidim/action_delegator/admin/manage_delegations/new.html.erb:14]: It's ok for me to make that change, it's easier to understand.

**Comment by microstudi** [comment]
@laurajaime you think you can rebase develop and fix the tests?
If not we will do it ourselves, no problem. Just let us know if you can do it

**Comment by laurajaime** [comment]
Hello @microstudi ,

Of course!! I have local changes  with some refactor. I hope soon to be able to push it.

**Comment by laurajaime** [comment]
@microstudi can you review it again?

I have refactored some things and fixed the tests.

**Comment by laurajaime** [comment]
I'm reviewing failing tests.

**Comment by laurajaime** [comment]
@microstudi the tests that failed are also failing in `develop` branch. All tests related to that feature are passing.

**Comment by microstudi** [comment]
you're right @laurajaime, we did some quick renaming of string that is causing this. I'm fixing it so you can rebase

**Review by microstudi — CHANGES_REQUESTED** [review]
Hi @laurajaime 
I've fixed the develop specs, can you rebase this PR?

Also, I have to ask you to remove the ca.yml and es.yml files because otherwise crodwin messes things up. Of course you can add it there later.

**Comment by laurajaime** [comment]
> Hi @laurajaime I've fixed the develop specs, can you rebase this PR?
> 
> Also, I have to ask you to remove the ca.yml and es.yml files because otherwise crodwin messes things up. Of course you can add it there later.

Hi  @microstudi , done!

**Review by microstudi — CHANGES_REQUESTED** [review]
Hi @laurajaime 

First of all nice job and thanks for the patience.
I've tested this with a random csv and it surprise me that was failing, didn't know why because the error message is misleading, I received this:
image

As you see it says "Granteer cannot be empty, etc", but that's not true, is not empty, the real reason is that the user is not registered in Decidim. So the message should state this instead of saying that is empty.

It would be also nice to deal with non-registered users the way we are dealing when importing participants, but we can do this in another issue/PR, see how is the participants list (all users are imported and you can see who is and who is not registered and invite them).
image

**Comment by laurajaime** [comment]
> As you see it says "Granteer cannot be empty, etc", but that's not true, is not empty, the real reason is that the user is not registered in Decidim. So the message should state this instead of saying that is empty.

You're right! I will look for an improvement of the error messages.

**Comment by microstudi** [comment]
@laurajaime I'm sorry, we need to finish this as soon as possible, I've merged and solved the conflicts in here https://github.com/coopdevs/decidim-module-action_delegator/commits/develop. Also maintained your contribution. Closing this PR now. Thanks for your amazing work!

**Comment by tramuntanal** [comment]
Ok @microstudi let us know if you need something from our side

**Pull request closed without merging**